#163388142 Users should be able to follow each other

PaulinhoNdegwa opened the pull request:
**What does this PR do?**
Creates _friends_ application to handle following and unfollowing users

**Description of tasks to be completed**

- Write test for follow and unfollow functionality
- Create a new app 'friends' in project
- Create friends model and following serializers
- Create view for following, unfollowing and counting followers

**Any background context you want to provide?**

* A user should be able to follow and be followed by other users
* A user can unfollow and be unfollowed by other users
* A user can view his/her followers (or other users followers)
* A user can see who they are following (or who other users are following)

**How should this be manually tested?**

* `git clone https://github.com/andela/ah-the-phoenix`
* `cd ah-the-phoenix`
* `git checkkout ft-user-following-163388142`
* `python3 -m venv virtual_env_name` to create a virtual environment and activate
* Add a `.env` file with all required environment variables
* `source .env` to load your environment variables
* `pip install -r requirements.txt`
* Run the migrations `python manage.py makemigrations && python manage.py migrate`

On Postman, register 2 or more users via the endpoint `http://127.0.0.1:8000/api/v1/users/`.

- Input the authentication token in the authorization header

**Following** 
- Do a `POST` on the `http://127.0.0.1:8000/api/v1/profile/<username>/follow/` where `username` is the username of the user you want to follow.
![image](https://user-images.githubusercontent.com/42174387/52647712-44a4e300-2ef6-11e9-80a2-5d74a2a526b0.png)

**Unfollowing**
- Do a `DELETE` on the `http://127.0.0.1:8000/api/v1/profile/<username>/follow/` where `username` is the username of the user you want to unfollow.
![image](https://user-images.githubusercontent.com/42174387/52647721-4bcbf100-2ef6-11e9-8b1b-643d9477ef8c.png)

**Get follower and following list**
- - Do a `GET` on the `http://127.0.0.1:8000/api/v1/profile/<username>/following/` where `username` is the username of the user you want to view following count.
![image](https://user-images.githubusercontent.com/42174387/52647736-538b9580-2ef6-11e9-9451-69bcf18884f2.png)

**What are the relevant Pivotal Tracker Stories?**
#163388142

PaulinhoNdegwa committed:
feat(user-following): Allow users to follow each other

- Write test for follow and unfollow functionality
- Create a new app 'friends' in project
- Create friends model and following serializers
- Create view for following, unfollowing and counting followers

[Delivers #163388142]

dev-jey commented on authors/apps/friends/views.py:
```diff
@@ -0,0 +1,130 @@
+        try:
+            to_be_followed = User.objects.get(username=username)
+        except Exception:
+            return Response({
```
Hello Paul. Because its a `resource that can not be found on the` **url**, you could raise a `NotFound` Exception instead of returning a Response. Check it out here
https://www.django-rest-framework.org/api-guide/exceptions/

PaulinhoNdegwa replied:
On it

PaulinhoNdegwa replied:
@codact I have researched and settled on presenting custom error messages that non-technical users can understand.

dev-jey replied:
ok..that will still work fine

sirghiny commented:
I don't think we need an entire application just for the followers. You can accomplish this task by adding fields to the user model to cater for following and followers.
Refer:
https://docs.djangoproject.com/en/dev/ref/models/fields/#django.db.models.ManyToManyField.symmetrical
https://stackoverflow.com/questions/11721157/django-many-to-many-m2m-relation-to-same-model

PaulinhoNdegwa committed:
feat(follow/unfollow): Implement follow/unfollow in user model

- Add a following(manytomanyfield) field in the User model
- Move follow/unfollow serializers and views to the auth app
- Refactor follow/unfollow code to utilize user model

[Finishes #163388142
]

PaulinhoNdegwa commented:
> I don't think we need an entire application just for the followers. You can accomplish this task by adding fields to the user model to cater for following and followers.
> Refer:
> https://docs.djangoproject.com/en/dev/ref/models/fields/#django.db.models.ManyToManyField.symmetrical
> https://stackoverflow.com/questions/11721157/django-many-to-many-m2m-relation-to-same-model

I redid that follow/unfollow feature in the `User` model. This method is way efficient. Thanks

sirghiny commented on authors/apps/authentication/models.py:
```diff
@@ -132,11 +136,11 @@ def get_short_name(self):
             "username": self.username,
-            "exp": int(time.strftime('%s'))
+            "exp": datetime.utcnow()
+            + timedelta(minutes=60*3)
```
You can simply have `timedelta(hours=3)`.

PaulinhoNdegwa replied:
Will change that

sirghiny commented on authors/apps/authentication/views.py:
```diff
@@ -276,3 +276,129 @@ def create(self, request):
+
+        if self.get_queryset() is not None:
+
+            followed_friend_objects = self.get_queryset()["followed"]
```
Assign a variable to the `queryset` so you don't have to query twice.

PaulinhoNdegwa replied:
Done

sirghiny commented on authors/apps/authentication/views.py:
```diff
@@ -276,3 +276,129 @@ def create(self, request):
+            users_followed = [u for u in followed_friend_objects]
+            users_following = [u for u in following_friend_objects]
+
+            follower_serializer = FollowerFollowingSerializer(
```
You can pass in the `querysets` from `get_queryset` with objects you don't have to create other lists.

PaulinhoNdegwa replied:
Done

PaulinhoNdegwa committed:
feat(follow/unfollow): Implement feedback

- Refactor redundant code
- Present follower's details after a successful follow request
- Restrict users from following unverified profiles

[Delivers #163388142]
